File: src/offer12.Path_in_Matrix/path_in_matrix.cpp
class Solution {
public:
    bool hasPath(char* matrix, int rows, int cols, char* str)
    {
        if(matrix == nullptr || rows == 0 || cols == 0 || str == nullptr) return nullptr;
        bool *mark = new bool[rows * cols];
        memset(mark, 0, rows*cols);
        int len = 0;
        for(int row = 0; row < rows; row++){
            for(int col = 0; col < cols; col++){
                if(hasPathCore(matrix, rows, cols, str, row, col, len, mark))
                    return true;
            }
        }
        delete[] mark;
        return false;
    }
    bool hasPathCore(char* matrix, int rows, int cols, char* str, int row, int col, int len, bool* mark){
        if(str[len] == '\0') return true;
        bool hasPath = false;
        if(row >= 0 && col >= 0 && row < rows && col < cols
          && matrix[cols * row + col] == str[len]
          && !mark[row * cols + col]){
            len++;
            mark[row * cols + col] = true;
            hasPath = hasPathCore(matrix, rows, cols, str, row, col-1, len, mark)
                || hasPathCore(matrix, rows, cols, str, row, col+1, len, mark)
                || hasPathCore(matrix, rows, cols, str, row-1, col, len, mark)
                || hasPathCore(matrix, rows, cols, str, row+1, col, len, mark);
            if(!hasPath){
                len--;
                mark[row*cols+col] = false;
            }
        }
        return hasPath;
    }

};

File: src/offer12.Path_in_Matrix/path_in_matrix.java
public class Solution {
    public boolean hasPath(char[] matrix, int rows, int cols, char[] str)
    {
        boolean[] visited = new boolean[matrix.length];
        int idx = 0;
        for(int i = 0; i < rows; i++){
            for(int j = 0; j < cols; j++){
                if(check(matrix, rows, cols, str, i, j, idx, visited)){
                    return true;
                }
            }
        }
        return false;
    }
    private boolean check(char[] matrix, int rows, int cols, char[] str, 
                          int row, int col, int idx, boolean[] visited){
        if(idx == str.length){
            return true;
        }
        boolean flag = false;
        if(row >= 0 && row < rows && col >= 0 && col < cols
              && !visited[row * cols + col]
              && matrix[row * cols + col] == str[idx]){
            idx++;
            visited[row * cols + col] = true;
            flag = check(matrix, rows, cols, str, row-1, col, idx, visited)
                    || check(matrix, rows, cols, str, row, col+1, idx, visited)
                    || check(matrix, rows, cols, str, row+1, col, idx, visited)
                    || check(matrix, rows, cols, str, row, col-1, idx, visited);
            if(!flag){
                idx--;
                visited[row * cols + col] = false;
            }
        }
        return flag;
    }
}